// @rc-component/menu/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Divider", {
  enumerable: true,
  get: function () {
    return _Divider.default;
  }
});
Object.defineProperty(exports, "Item", {
  enumerable: true,
  get: function () {
    return _MenuItem.default;
  }
});
Object.defineProperty(exports, "ItemGroup", {
  enumerable: true,
  get: function () {
    return _MenuItemGroup.default;
  }
});
Object.defineProperty(exports, "MenuItem", {
  enumerable: true,
  get: function () {
    return _MenuItem.default;
  }
});
Object.defineProperty(exports, "MenuItemGroup", {
  enumerable: true,
  get: function () {
    return _MenuItemGroup.default;
  }
});
Object.defineProperty(exports, "SubMenu", {
  enumerable: true,
  get: function () {
    return _SubMenu.default;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "useFullPath", {
  enumerable: true,
  get: function () {
    return _PathContext.useFullPath;
  }
});
var _Menu = _interopRequireDefault(require("./Menu"));
var _MenuItem = _interopRequireDefault(require("./MenuItem"));
var _SubMenu = _interopRequireDefault(require("./SubMenu"));
var _MenuItemGroup = _interopRequireDefault(require("./MenuItemGroup"));
var _PathContext = require("./context/PathContext");
var _Divider = _interopRequireDefault(require("./Divider"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const ExportMenu = _Menu.default;
ExportMenu.Item = _MenuItem.default;
ExportMenu.SubMenu = _SubMenu.default;
ExportMenu.ItemGroup = _MenuItemGroup.default;
ExportMenu.Divider = _Divider.default;
var _default = exports.default = ExportMenu;
